File: wrappers/csharp/LibRangeMap/SequenceRangeMapper.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class SequenceRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "sequence_range";

    private readonly IntegerRangeMapper? _integerMapper;
    private readonly FloatRangeMapper? _floatMapper;
    private readonly BooleanRangeMapper? _booleanMapper;
    private readonly BytesRangeMapper? _bytesMapper;
    private readonly CategoricalRangeMapper? _categoricalMapper;

    public string ElementMapperType { get; }
    public bool AllowEmpty { get; }
    public string? Name { get; }

    public SequenceRangeMapper(IntegerRangeMapper elementMapper, bool allowEmpty = false, string? name = null)
    {
        _integerMapper = elementMapper ?? throw new ArgumentNullException(nameof(elementMapper));
        ElementMapperType = "integer_range";
        AllowEmpty = allowEmpty;
        Name = NormalizeName(name);
    }

    public SequenceRangeMapper(FloatRangeMapper elementMapper, bool allowEmpty = false, string? name = null)
    {
        _floatMapper = elementMapper ?? throw new ArgumentNullException(nameof(elementMapper));
        ElementMapperType = "float_range";
        AllowEmpty = allowEmpty;
        Name = NormalizeName(name);
    }

    public SequenceRangeMapper(BooleanRangeMapper elementMapper, bool allowEmpty = false, string? name = null)
    {
        _booleanMapper = elementMapper ?? throw new ArgumentNullException(nameof(elementMapper));
        ElementMapperType = "boolean_range";
        AllowEmpty = allowEmpty;
        Name = NormalizeName(name);
    }

    public SequenceRangeMapper(BytesRangeMapper elementMapper, bool allowEmpty = false, string? name = null)
    {
        _bytesMapper = elementMapper ?? throw new ArgumentNullException(nameof(elementMapper));
        ElementMapperType = "bytes_range";
        AllowEmpty = allowEmpty;
        Name = NormalizeName(name);
    }

    public SequenceRangeMapper(CategoricalRangeMapper elementMapper, bool allowEmpty = false, string? name = null)
    {
        _categoricalMapper = elementMapper ?? throw new ArgumentNullException(nameof(elementMapper));
        ElementMapperType = "categorical_range";
        AllowEmpty = allowEmpty;
        Name = NormalizeName(name);
    }

    public double[] MapValues(IEnumerable<long> values)
    {
        EnsureIntegerMapper();
        if (values is null)
        {
            throw new ArgumentNullException(nameof(values), "sequence payload must not be null.");
        }

        if (!AllowEmpty && !values.Any())
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return MapValues(values, value => _integerMapper!.MapValue(value));
    }

    public double[] MapValues(IEnumerable<double> values)
    {
        EnsureFloatMapper();
        if (values is null)
        {
            throw new ArgumentNullException(nameof(values), "sequence payload must not be null.");
        }

        if (!AllowEmpty && !values.Any())
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return MapValues(values, value => _floatMapper!.MapValue(value));
    }

    public double[] MapValues(IEnumerable<bool> values)
    {
        EnsureBooleanMapper();
        if (values is null)
        {
            throw new ArgumentNullException(nameof(values), "sequence payload must not be null.");
        }

        if (!AllowEmpty && !values.Any())
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return MapValues(values, value => _booleanMapper!.MapValue(value));
    }

    public double[] MapValues(IEnumerable<byte> values)
    {
        EnsureBytesMapper();
        if (values is null)
        {
            throw new ArgumentNullException(nameof(values), "sequence payload must not be null.");
        }

        if (!AllowEmpty && !values.Any())
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return MapValues(values, value => _bytesMapper!.MapValue(value));
    }

    public double[] MapValues(IEnumerable<object?> values)
    {
        EnsureCategoricalMapper();
        if (values is null)
        {
            throw new ArgumentNullException(nameof(values), "sequence payload must not be null.");
        }

        if (!AllowEmpty && !values.Any())
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return MapValues(values, value => _categoricalMapper!.MapValue(value));
    }

    public object MapNested(object value)
    {
        return MapNestedRecursive(value);
    }

    public object Map(IEnumerable value)
    {
        return MapNested(value);
    }

    public string ToJson()
    {
        return JsonSerializer.Serialize(ToSpec(), JsonOptions);
    }

    public static SequenceRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<SequenceMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static SequenceRangeMapper FromSpec(SequenceMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }
        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }
        if (spec.ElementMapperType is null)
        {
            throw new ArgumentException("element_mapper_type is required.", nameof(spec));
        }
        if (spec.ElementMapperSpec is null)
        {
            throw new ArgumentException("element_mapper_spec is required.", nameof(spec));
        }

        return spec.ElementMapperType switch
        {
            "integer_range" => new SequenceRangeMapper(IntegerRangeMapper.FromJson(spec.ElementMapperSpec.Value.GetRawText()), spec.AllowEmpty, spec.Name),
            "float_range" => new SequenceRangeMapper(FloatRangeMapper.FromJson(spec.ElementMapperSpec.Value.GetRawText()), spec.AllowEmpty, spec.Name),
            "boolean_range" => new SequenceRangeMapper(BooleanRangeMapper.FromJson(spec.ElementMapperSpec.Value.GetRawText()), spec.AllowEmpty, spec.Name),
            "bytes_range" => new SequenceRangeMapper(BytesRangeMapper.FromJson(spec.ElementMapperSpec.Value.GetRawText()), spec.AllowEmpty, spec.Name),
            "categorical_range" => new SequenceRangeMapper(CategoricalRangeMapper.FromJson(spec.ElementMapperSpec.Value.GetRawText()), spec.AllowEmpty, spec.Name),
            _ => throw new ArgumentException($"unsupported element_mapper_type '{spec.ElementMapperType}'.", nameof(spec))
        };
    }

    private object MapNestedRecursive(object value)
    {
        if (value is null)
        {
            throw new ArgumentNullException(nameof(value), "sequence payload must not be null.");
        }

        if (value is string)
        {
            throw new ArgumentException("string payload is not supported for sequence mapping. use text mapper or bytes mapping.");
        }

        if (IsScalarValue(value))
        {
            return MapScalar(value);
        }

        if (value is IDictionary)
        {
            throw new ArgumentException("dict-like values are not supported for sequence mapping. Use map/object family.");
        }

        if (value is IEnumerable enumerable)
        {
            return MapSequence(enumerable);
        }

        throw new InvalidOperationException($"unsupported value type '{value.GetType().FullName}'.");
    }

    private object MapSequence(IEnumerable sequence)
    {
        var mapped = new List<object>();
        foreach (var item in sequence)
        {
            mapped.Add(MapNestedRecursive(item));
        }

        if (mapped.Count == 0 && !AllowEmpty)
        {
            throw new ArgumentException("sequence payload must not be empty unless allow_empty is true.");
        }

        return mapped.ToArray();
    }

    private double MapScalar(object value)
    {
        return ElementMapperType switch
        {
            "integer_range" => _integerMapper!.MapValue(ConvertToLong(value)),
            "float_range" => _floatMapper!.MapValue(ConvertToDouble(value)),
            "boolean_range" => _booleanMapper!.MapValue(ConvertToBoolean(value)),
            "bytes_range" => _bytesMapper!.MapValue(ConvertToByte(value)),
            "categorical_range" => _categoricalMapper!.MapValue(value),
            _ => throw new InvalidOperationException($"invalid sequence element mapper '{ElementMapperType}'.")
        };
    }

    private static double[] MapValues<T>(IEnumerable<T> values, Func<T, double> mapValue)
    {
        return values.Select(mapValue).ToArray();
    }

    private static long ConvertToLong(object value)
    {
        return value switch
        {
            sbyte v => v,
            byte v => v,
            short v => v,
            ushort v => v,
            int v => v,
            uint v => checked((long)v),
            long v => v,
            ulong v => checked((long)v),
            _ => throw new ArgumentException("integer sequence expects integer values only.")
        };
    }

    private static double ConvertToDouble(object value)
    {
        return value switch
        {
            float v => v,
            double v => v,
            decimal v => (double)v,
            _ => throw new ArgumentException("float sequence expects float values only.")
        };
    }

    private static bool ConvertToBoolean(object value)
    {
        return value switch
        {
            bool v => v,
            _ => throw new ArgumentException("boolean sequence expects bool values only.")
        };
    }

    private static long ConvertToByte(object value)
    {
        return value is byte v
            ? v
            : throw new ArgumentException("bytes sequence expects byte values only.");
    }

    private static bool IsScalarValue(object value) =>
        value switch
        {
            string or byte[] => false,
            char or bool or sbyte or byte or short or ushort or int or uint or long or ulong or float or double or decimal => true,
            null => false,
            _ => value is not IEnumerable
        };

    private SequenceMapperSpec ToSpec()
    {
        var elementSpec = ElementMapperType switch
        {
            "integer_range" => _integerMapper!.ToJson(),
            "float_range" => _floatMapper!.ToJson(),
            "boolean_range" => _booleanMapper!.ToJson(),
            "bytes_range" => _bytesMapper!.ToJson(),
            "categorical_range" => _categoricalMapper!.ToJson(),
            _ => throw new InvalidOperationException($"invalid element mapper '{ElementMapperType}'.")
        };

        if (elementSpec is not string elementSpecJson)
        {
            throw new InvalidOperationException("Unable to serialize element mapper.");
        }

        return new SequenceMapperSpec
        {
            SpecVersion = SpecVersion,
            MapperType = SpecType,
            ElementMapperType = ElementMapperType,
            ElementMapperSpec = JsonDocument.Parse(elementSpecJson).RootElement.Clone(),
            AllowEmpty = AllowEmpty,
            Name = Name
        };
    }

    private static string? NormalizeName(string? name)
    {
        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }

        return name;
    }

    private void EnsureIntegerMapper()
    {
        if (_integerMapper is null)
        {
            throw new InvalidOperationException("sequence mapper is not configured for integer values.");
        }
    }

    private void EnsureFloatMapper()
    {
        if (_floatMapper is null)
        {
            throw new InvalidOperationException("sequence mapper is not configured for float values.");
        }
    }

    private void EnsureBooleanMapper()
    {
        if (_booleanMapper is null)
        {
            throw new InvalidOperationException("sequence mapper is not configured for boolean values.");
        }
    }

    private void EnsureBytesMapper()
    {
        if (_bytesMapper is null)
        {
            throw new InvalidOperationException("sequence mapper is not configured for bytes values.");
        }
    }

    private void EnsureCategoricalMapper()
    {
        if (_categoricalMapper is null)
        {
            throw new InvalidOperationException("sequence mapper is not configured for categorical values.");
        }
    }
}

public sealed class SequenceMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "sequence_range";

    [JsonPropertyName("element_mapper_type")]
    public string? ElementMapperType { get; init; }

    [JsonPropertyName("element_mapper_spec")]
    public JsonElement? ElementMapperSpec { get; init; }

    [JsonPropertyName("allow_empty")]
    public bool AllowEmpty { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}

File: wrappers/java/src/librangemap/SequenceRangeMapper.java
package librangemap;

import java.util.ArrayList;
import java.util.List;

public final class SequenceRangeMapper {
    private static final String SPEC_VERSION = "1.0-alpha";
    private static final String MAPPER_TYPE = "sequence_range";

    private final Object elementMapper;
    private final boolean allowEmpty;
    private final String name;

    public SequenceRangeMapper(Object elementMapper) {
        this(elementMapper, false, null);
    }

    public SequenceRangeMapper(Object elementMapper, boolean allowEmpty) {
        this(elementMapper, allowEmpty, null);
    }

    public SequenceRangeMapper(Object elementMapper, boolean allowEmpty, String name) {
        if (elementMapper == null) {
            throw new IllegalArgumentException("elementMapper must not be null.");
        }
        validateElementMapper(elementMapper);
        if (name != null && name.isEmpty()) {
            throw new IllegalArgumentException("name must not be empty");
        }
        this.elementMapper = elementMapper;
        this.allowEmpty = allowEmpty;
        this.name = name;
    }

    public Object mapValue(Object value) {
        if (value == null) {
            throw new IllegalArgumentException("value must not be null.");
        }

        if (value instanceof List<?>) {
            List<?> items = (List<?>) value;
            if (!allowEmpty && items.isEmpty()) {
                throw new IllegalArgumentException(
                        "empty sequence is invalid by default; set allowEmpty=true to map empty containers."
                );
            }
            List<Object> mapped = new ArrayList<>(items.size());
            for (Object item : items) {
                mapped.add(mapValueElement(item));
            }
            return mapped;
        }

        if (value.getClass().isArray()) {
            int length = java.lang.reflect.Array.getLength(value);
            if (!allowEmpty && length == 0) {
                throw new IllegalArgumentException(
                        "empty sequence is invalid by default; set allowEmpty=true to map empty containers."
                );
            }
            Object[] mapped = new Object[length];
            for (int i = 0; i < length; i++) {
                mapped[i] = mapValueElement(java.lang.reflect.Array.get(value, i));
            }
            return mapped;
        }

        return mapScalar(value);
    }

    private Object mapValueElement(Object value) {
        if (elementMapper instanceof CategoricalRangeMapper categoricalMapper) {
            return categoricalMapper.mapValue(value);
        }
        if (elementMapper instanceof TextRangeMapper textMapper) {
            if (!(value instanceof String)) {
                return mapScalar(value);
            }
            return textMapper.mapValue((String) value);
        }
        if (elementMapper instanceof BytesRangeMapper bytesMapper) {
            if (value instanceof String || value instanceof byte[] || value instanceof int[]) {
                return bytesMapper.mapValue(value);
            }
            return mapScalar(value);
        }
        if (elementMapper instanceof SequenceRangeMapper nestedMapper) {
            return nestedMapper.mapValue(value);
        }
        if (elementMapper instanceof ObjectRangeMapper objectMapper) {
            return objectMapper.mapValue(value);
        }
        if (elementMapper instanceof TemporalRangeMapper temporalMapper) {
            return temporalMapper.mapValue(value);
        }
        if (value == null) {
            return mapScalar(value);
        }
        Class<?> valueType = value.getClass();
        if (valueType.isArray()) {
            return mapValue(value);
        }
        if (value instanceof List<?>) {
            return mapValue(value);
        }
        return mapScalar(value);
    }

    public Object map(Object value) {
        return mapValue(value);
    }

    public String toJson() {
        StringBuilder sb = new StringBuilder();
        sb.append('{');
        sb.append("\"spec_version\":\"").append(SPEC_VERSION).append("\",");
        sb.append("\"mapper_type\":\"").append(MAPPER_TYPE).append("\",");
        sb.append("\"element_mapper\":").append(mapperToJson(elementMapper)).append(',');
        sb.append("\"allow_empty\":").append(allowEmpty);
        if (name != null) {
            sb.append(",\"name\":\"").append(escape(name)).append("\"");
        }
        sb.append('}');
        return sb.toString();
    }

    public static SequenceRangeMapper fromJson(String json) {
        if (json == null) {
            throw new IllegalArgumentException("mapper json must not be null");
        }

        String specVersion = extractString(json, "spec_version");
        if (!SPEC_VERSION.equals(specVersion)) {
            throw new IllegalArgumentException("unsupported spec_version " + specVersion);
        }

        String mapperType = extractString(json, "mapper_type");
        if (!MAPPER_TYPE.equals(mapperType)) {
            throw new IllegalArgumentException("unsupported mapper_type " + mapperType);
        }

        String elementMapperText = extractObject(json, "element_mapper");
        Object elementMapper = mapperFromJson(elementMapperText);
        Boolean allowEmpty = extractBoolean(json, "allow_empty");
        String name = extractStringOrNull(json, "name");

        return new SequenceRangeMapper(elementMapper, allowEmpty != null && allowEmpty, name);
    }

    private Object mapScalar(Object value) {
        if (elementMapper instanceof IntegerRangeMapper integerMapper) {
            if (!(value instanceof Number)) {
                throw new IllegalArgumentException("value must be a finite integer.");
            }
            return integerMapper.mapValue(((Number) value).longValue());
        }

        if (elementMapper instanceof FloatRangeMapper floatMapper) {
            if (!(value instanceof Number)) {
                throw new IllegalArgumentException("value must be a finite number.");
            }
            return floatMapper.mapValue(((Number) value).doubleValue());
        }

        if (elementMapper instanceof BooleanRangeMapper boolMapper) {
            if (!(value instanceof Boolean)) {
                throw new IllegalArgumentException("value must be a boolean.");
            }
            return boolMapper.mapValue((Boolean) value);
        }
        if (elementMapper instanceof CategoricalRangeMapper categoricalMapper) {
            return categoricalMapper.mapValue(value);
        }

        if (elementMapper instanceof TextRangeMapper textMapper) {
            if (!(value instanceof String)) {
                throw new IllegalArgumentException("value must be a string.");
            }
            return textMapper.mapValue((String) value);
        }

        if (elementMapper instanceof BytesRangeMapper bytesMapper) {
            if (value instanceof byte[] bytes) {
                return bytesMapper.mapValue(bytes);
            }
            if (value instanceof int[] bytes) {
                return bytesMapper.mapValue(bytes);
            }
            if (value instanceof String text) {
                return bytesMapper.mapValue(text);
            }
            throw new IllegalArgumentException("value must be a byte[] int[] or string.");
        }

        if (elementMapper instanceof SequenceRangeMapper sequenceMapper) {
            return sequenceMapper.mapValue(value);
        }

        if (elementMapper instanceof TemporalRangeMapper temporalMapper) {
            return temporalMapper.mapValue(value);
        }

        throw new IllegalArgumentException("unsupported element mapper type " + elementMapper.getClass().getName());
    }

    private static Object mapperFromJson(String json) {
        String mapperType = extractString(json, "mapper_type");
        return switch (mapperType) {
            case "integer_range" -> IntegerRangeMapper.fromJson(json);
            case "float_range" -> FloatRangeMapper.fromJson(json);
            case "boolean_range" -> BooleanRangeMapper.fromJson(json);
            case "categorical_range" -> CategoricalRangeMapper.fromJson(json);
            case "text_range" -> TextRangeMapper.fromJson(json);
            case "bytes_range" -> BytesRangeMapper.fromJson(json);
            case "sequence_range" -> SequenceRangeMapper.fromJson(json);
            case "map_range" -> ObjectRangeMapper.fromJson(json);
            case "temporal_range" -> TemporalRangeMapper.fromJson(json);
            default -> throw new IllegalArgumentException("unsupported mapper_type " + mapperType);
        };
    }

    private static String mapperToJson(Object mapper) {
        if (mapper instanceof IntegerRangeMapper integerMapper) {
            return integerMapper.toJson();
        }
        if (mapper instanceof FloatRangeMapper floatMapper) {
            return floatMapper.toJson();
        }
        if (mapper instanceof BooleanRangeMapper boolMapper) {
            return boolMapper.toJson();
        }
        if (mapper instanceof TextRangeMapper textMapper) {
            return textMapper.toJson();
        }
        if (mapper instanceof BytesRangeMapper bytesMapper) {
            return bytesMapper.toJson();
        }
        if (mapper instanceof CategoricalRangeMapper categoricalMapper) {
            return categoricalMapper.toJson();
        }
        if (mapper instanceof SequenceRangeMapper sequenceMapper) {
            return sequenceMapper.toJson();
        }
        if (mapper instanceof ObjectRangeMapper objectMapper) {
            return objectMapper.toJson();
        }
        if (mapper instanceof TemporalRangeMapper temporalMapper) {
            return temporalMapper.toJson();
        }
        throw new IllegalArgumentException("unsupported element mapper type " + mapper.getClass().getName());
    }

    private static void validateElementMapper(Object elementMapper) {
        if (elementMapper instanceof IntegerRangeMapper
                || elementMapper instanceof FloatRangeMapper
                || elementMapper instanceof BooleanRangeMapper
                || elementMapper instanceof CategoricalRangeMapper
                || elementMapper instanceof TextRangeMapper
                || elementMapper instanceof BytesRangeMapper
                || elementMapper instanceof SequenceRangeMapper
                || elementMapper instanceof ObjectRangeMapper
                || elementMapper instanceof TemporalRangeMapper) {
            return;
        }
        throw new IllegalArgumentException("elementMapper must be a recognized mapper instance.");
    }

    private static String extractObject(String json, String key) {
        String needle = "\"" + key + "\":";
        int start = json.indexOf(needle);
        if (start < 0) {
            throw new IllegalArgumentException("missing key " + key);
        }
        start += needle.length();
        while (start < json.length() && Character.isWhitespace(json.charAt(start))) {
            start++;
        }
        if (start >= json.length() || json.charAt(start) != '{') {
            throw new IllegalArgumentException("expected object for " + key);
        }

        int depth = 0;
        boolean inString = false;
        boolean escape = false;
        int end = start;
        for (; end < json.length(); end++) {
            char c = json.charAt(end);
            if (escape) {
                escape = false;
                continue;
            }
            if (c == '\\') {
                escape = true;
                continue;
            }
            if (c == '"') {
                inString = !inString;
                continue;
            }
            if (inString) {
                continue;
            }
            if (c == '{') {
                depth++;
            } else if (c == '}') {
                depth--;
                if (depth == 0) {
                    return json.substring(start, end + 1);
                }
            }
        }
        throw new IllegalArgumentException("unterminated object for " + key);
    }

    private static String extractString(String text, String key) {
        String needle = "\"" + key + "\":\"";
        int start = text.indexOf(needle);
        if (start < 0) {
            throw new IllegalArgumentException("missing key " + key);
        }
        start += needle.length();
        int end = text.indexOf('"', start);
        if (end < 0) {
            throw new IllegalArgumentException("unterminated string for " + key);
        }
        return text.substring(start, end);
    }

    private static String extractStringOrNull(String text, String key) {
        String needle = "\"" + key + "\":\"";
        int start = text.indexOf(needle);
        if (start < 0) {
            return null;
        }
        start += needle.length();
        int end = text.indexOf('"', start);
        if (end < 0) {
            return null;
        }
        return text.substring(start, end);
    }

    private static Boolean extractBoolean(String text, String key) {
        String needle = "\"" + key + "\":";
        int start = text.indexOf(needle);
        if (start < 0) {
            return null;
        }
        start += needle.length();
        if (text.startsWith("true", start)) {
            return Boolean.TRUE;
        }
        if (text.startsWith("false", start)) {
            return Boolean.FALSE;
        }
        throw new IllegalArgumentException("invalid boolean for " + key);
    }

    private static String escape(String text) {
        return text.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
